### PR TITLE
[processor/logdeduplicationprocessor] Add logdedupprocessor

### LINKED ISSUE #34118: New component: Log deduplication processor
### The purpose and use-cases of the new component

We're working to put together a log de-duplication processor at Honeycomb and wanted to get the temperature of the community for hosting such a component in the contrib repository.

### Example configuration for the component

```yaml
dedupe:
  // TTL is the time-to-live for each entry in the cache. Default is 30 seconds.
  ttl: 30s
  // MaxEntries is the maximum number of entries that can be stored in the cache. Default is 1000.
  max_entries: 1000
  // AttributeNames is the list of attributes to use as the key for deduplication.
  attribute_names: []
```

### Telemetry data types supported

logs

### Is this a vendor-specific component?

- [ ] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am a member of the OpenTelemetry organization.
- [ ] If this is a vendor-specific component, I am proposing to contribute and support it as a representative of the vendor.

### Code Owner(s)

@codeboten @MikeGoldsmith

### Sponsor (optional)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
**Description:**

Starts the donation of the [logdedupprocessor](https://github.com/observIQ/bindplane-agent/tree/release/v1.58.0/processor/logdeduplicationprocessor) from ObserveIQ's Bindplane agent on behalf of @BinaryFissionGames.

**Link to tracking Issue:**

- Closes #34118 

**Testing:**

Includes unit tests.

**Documentation:**

### CODE REVIEW

File: .github/CODEOWNERS
Comment (BinaryFissionGames) [inline .github/CODEOWNERS:177]: I think this file is supposed to be generated from some other file, maybe from a `metadata.yaml` file for mdatagen?

I think we'll want that either way, we'll want to mark this component as in development while it's being tweaked (I think we'll at least want a limit on the max entries like the original proposal before putting this in alpha).

File: processor/logdeduplicationprocessor/timezone.md
Comment (BinaryFissionGames) [inline processor/logdeduplicationprocessor/timezone.md:1]: I'm thinking we can remove this file, I think it's acceptable in this PR or in future ones.

File: .github/CODEOWNERS
Comment (MikeGoldsmith) [reply .github/CODEOWNERS:177]: mdatagen can help with managing stability, capabilities and any internal telemetry (traces / metrics) we want to add. I can add that here, I was trying to keep this as clean and true of your work.

File: processor/logdeduplicationprocessor/timezone.md
Comment (MikeGoldsmith) [reply processor/logdeduplicationprocessor/timezone.md:1]: Do you think we should remove the link, or replace it with something else?

File: .github/CODEOWNERS
Comment (BinaryFissionGames) [reply .github/CODEOWNERS:177]: I think we'll want to add it at the very least for the stability, so anyone who comes across this component understands it's still being tweaked and could have breaking changes introduced.

File: processor/logdeduplicationprocessor/README.md
Comment (BinaryFissionGames) [inline processor/logdeduplicationprocessor/README.md:23]: For removing timezone.md, I think we could replace the link like this:
```suggestion
| timezone     | string | `UTC`    | The timezone of the `first_observed_timestamp` and `last_observed_timestamp` timestamps on the emitted aggregated log. The available locations depend on the local IANA Time Zone database. [This page](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) contains many examples, such as `America/New_York`.  |
```

The description is adapted from the stanza docs for a similar field: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/docs/types/timestamp.md

File: processor/logdeduplicationprocessor/timezone.md
Comment (BinaryFissionGames) [reply processor/logdeduplicationprocessor/timezone.md:1]: Added a suggestion for what to do about the link.

File: processor/logdeduplicationprocessor/field_remover.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/field_remover.go:16]: I assume we'll want to update the field syntax to match OTTL. I think it's ok to leave for now as long as there's a tracking issue and the component is marked as in development.

File: processor/logdeduplicationprocessor/processor.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/processor.go:48]: Can we use the context passed in?

File: processor/logdeduplicationprocessor/processor.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/processor.go:79]: Is all of this extraneous? Create a local done channel, listen for it to close, but defer closing it?

The only thing somewhat useful this might be doing is returning `ctx.Err()` in the case where the passed in context is canceled while we're shutting down. I'm not convinced that's useful though and not aware of any other cases where we do that in `Shutdown`.

File: .github/CODEOWNERS
Comment (MikeGoldsmith) [reply .github/CODEOWNERS]: Added metadata and ran mdatagen in [ece5fba](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34465/commits/ece5fba3960e47383d67f9e54b69cea6581bcf93)

File: processor/logdeduplicationprocessor/counter.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/counter.go:121]: nit: If `newLogCounter` handles the timestamp
```suggestion
		r.logCounters[key] = newLogCounter(logRecord)
```

File: processor/logdeduplicationprocessor/metadata.yaml
Comment (BinaryFissionGames) [inline processor/logdeduplicationprocessor/metadata.yaml:10]: We'll want to add @djaglowski as a codeowner as well, since he's the sponsor for the component.

File: processor/logdeduplicationprocessor/counter.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/counter.go:148]:
```suggestion
// getLogKey creates a unique hash for the log record to use as a map key
```
Since it appears we're using fnv

**Review by BinaryFissionGames — COMMENTED** [review]
This looks good to me as the initial pass, pinging @djaglowski to take a look.

Thanks for taking this on!

File: processor/logdeduplicationprocessor/counter.go
Comment (djaglowski) [inline processor/logdeduplicationprocessor/counter.go:85]: Since `Add` is typically called repeatedly for each resource, we should hash the resource attributes in the caller and pass them in. Especially important because all of this is within a mux.

**Review by djaglowski — COMMENTED** [review]
Generally looks good to me. I recognize this is a preliminary commit but I used the opportunity to give it a closer look. My feedback can be set aside for future PRs if there's a preference to have a cleaner copy of the original source.

**Comment by MikeGoldsmith** [comment]
> Generally looks good to me. I recognize this is a preliminary commit but I used the opportunity to give it a closer look. My feedback can be set aside for future PRs if there's a preference to have a cleaner copy of the original source.

I wanted to get a approval from @BinaryFissionGames at this PR is on their behalf. After that I also have some suggestions.

I'm happy to address your feedback either in this PR or in subsequent PRs so we can make more isolated changes.

File: processor/logdeduplicationprocessor/field_remover.go
Comment (MikeGoldsmith) [reply processor/logdeduplicationprocessor/field_remover.go:16]: Opened tracking issue:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34473

File: processor/logdeduplicationprocessor/processor.go
Comment (MikeGoldsmith) [reply processor/logdeduplicationprocessor/processor.go:79]: Opened tracking issue:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34478

**Comment by MikeGoldsmith** [comment]
@djaglowski I think I've addressed all the immediate feedback and creates tracking issues for the others:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34473
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34478

Feel free to assign these to me.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
